Model Discovery

marvinirwin opened the pull request:
Adds the command `lb4 discover` command and its documentation.

The command has a `schemaDefs` option (the path to a file which contains a schema definition, or array of lb3 schema definitions) which overrides everything and simply creates the corresponding typescript file.  It's used for the integration test, and could also solve #2224

It's a piece of #2090
Implements #1921

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

bajtos commented on packages/cli/lib/model-discoverer.js:
```diff
@@ -0,0 +1,111 @@
+function discoverModelNames(ds, options) {
+  return new Promise((resolve, reject) => {
+    function discover() {
+      ds.discoverModelDefinitions(options, (err, schemas) => {
```
Discovery APIs support Promises, no need to use callbacks.

```ts
const schemas = await ds.discoverModelDefinitions(options);
// etc.
```

Please rewrite all helpers in this file to use `await` instead of callbacks.

marvinirwin replied:
Done

bajtos commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,41 @@
+lang: en
+title: "Discovering models from relational databases"
+keywords: LoopBack
+sidebar: lb3_sidebar
```
Please fix to `lb4_sidebar`.

This new file needs to be added to the sidebar, see https://github.com/strongloop/loopback-next/blob/master/docs/site/sidebars/lb4_sidebar.yml

marvinirwin replied:
Done

bajtos commented on packages/cli/generators/model/templates/model.ts.ejs:
```diff
@@ -5,7 +5,7 @@ import {model, property} from '@loopback/repository';
 <% } -%>
 
-@model()
+@model(<%- locals.settings ? JSON.stringify({settings}) : '' %>)
```
Where is `locals.settings` coming from? Did you perhaps mean to check `settings`?

marvinirwin replied:
Yes, this was put in to check if the settings object existed.  Is there a more idiomatic way to do this?

bajtos replied:
Have you tried to following?

```suggestion
@model(<%- settings ? JSON.stringify({settings}) : '' %>)
```

marvinirwin replied:
It will throw a reference error if it's not there, I got my solution from here

https://stackoverflow.com/questions/5372559/what-is-the-proper-way-to-check-for-existence-of-variable-in-an-ejs-template-us

raymondfeng replied:
`locals` is special variable for EJS and it's safer to use `locals.settings` if `settings` can be `undefined` when the EJS template is rendered.

bajtos replied:
Makes sense, I was not aware of this aspect of EJS.

Thank you for being patient with me and explaining the details, I learned something new today 😄

bajtos commented on packages/cli/test/fixtures/discover/index.js:
```diff
@@ -0,0 +1,23 @@
+    path: '',
+    file: 'test-schema-defs.json',
+    content: fs.readFileSync(
+      require.resolve('./models/test-schema-defs.json.txt'),
```
I am confused, why are you using `.json.txt` extension? Can we store JSON content in files with `.json` extension please?

marvinirwin replied:
Done

bajtos commented on packages/cli/test/fixtures/discover/index.js:
```diff
@@ -0,0 +1,23 @@
+    path: 'dist/src/datasources',
+    file: 'mem.datasource.json',
+    content: fs.readFileSync(
+      require.resolve('./models/mem.datasource.json.txt'),
```
Ditto, can we use `mem.datasource.json`?

marvinirwin replied:
Done

bajtos commented on packages/cli/test/fixtures/discover/index.js:
```diff
@@ -0,0 +1,23 @@
+  {
+    path: 'dist/src/datasources',
+    file: 'mem.datasource.js',
+    content: fs.readFileSync(require.resolve('./models/mem.datasource.js.txt')),
```
And similarly here, can we use `mem.datasource.js` please?

marvinirwin replied:
Changing this extension to `.js` causes something to try and load the file, which then gets a `Cannot find module '@loopback/core'` error when I run `npm test`.  

The reason I had them all with the `.txt` extension is because they did it in `fixtures/repositories/models/`, is avoiding this error why?

bajtos replied:
Fair enough, let's follow the current convention of using `.txt` files for now. Dealing with issues caused by using a different extension is out of scope of this pull request, I think.

raymondfeng commented on packages/cli/generators/discover/index.js:
```diff
@@ -0,0 +1,292 @@
+    this.option('discoverAll', {
+      type: Boolean,
+      description:
+        'Avoids the prompt asking for specific models and just discovers them all',
```
Is `Discover all models without prompting users to select` a better description?

marvinirwin replied:
Yes, that's much better.

raymondfeng commented on packages/cli/generators/discover/index.js:
```diff
@@ -0,0 +1,292 @@
+    if (this.shouldExit()) return;
+    const prompts = [
+      {
+        name: 'generateRepository',
```
The name should be `generateController`?

raymondfeng replied:
Where are these answers being used? I don't see code to generate controller/repository classes based on this prompt.

marvinirwin replied:
Fixed!  I had briefly toyed with entire api generation.

raymondfeng commented:
@marvinirwin Great effort! Please see my comments.

bajtos commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,41 @@
+
+`--views`: Choose whether to discover views, default true
+
+`-- discoverAll`: Skips the model prompt and discovers all of them
```
+1, I like `lb4 discover --all`

marvinirwin replied:
Done

marvinirwin commented on packages/cli/generators/discover/index.js:
```diff
@@ -0,0 +1,292 @@
+      type: String,
+      description:
+        'The path to a .json file with an array of schema definition',
+      default: undefined,
```
Yes.  Is there a way I can mark that fact for yeoman?

bajtos replied:
Probably not. Can you mention the exclusivity in `description` please?

marvinirwin commented:
@raymondfeng  @bajtos   Is there anything else I can do to help this get merged?  I think I've addressed all the reviewers' concerns.

dhmlau commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,43 @@
+## Overview
+Models can be discovered from a supported datasource by running the `lb4 discover` command.
+
+**The loopback project must be built and contain the built datasource files in `PROJECT_DIR/dist/src/datasources.js**
```
`loopback` -> `LoopBack`

marvinirwin replied:
Fixed!

dhmlau commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,43 @@
+
+`--dataSource`: Put a valid datasource name here to skip the datasource prompt
+
+`--views`: Choose whether to discover views, default true
```
`Choose whether to discover views, default true` -> `Choose whether to discover views.  Default is true`

marvinirwin replied:
Done

dhmlau commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,43 @@
+
+`--schema`: Specify the schema which the datasource will find the models to discover
+
+`--schemaDefs`: The path to a `.json` file containing a model definition, like those used loopback 2 and 3.  Used to make a typescript model out of an old loopback model without any prompts.
```
- `loopback` -> `LoopBack`
- For the second sentence, I'm wondering if it would be better:
   ```
   LoopBack 4 models will be created out of the LoopBack 3 models specified in `.json` file without any prompts.
   ```

marvinirwin replied:
Significantly better, fixed.

dhmlau commented on packages/cli/generators/discover/index.js:
```diff
@@ -0,0 +1,272 @@
+
+    this.option('schema', {
+      type: String,
+      description: 'Schema to discover, required for postgresql',
```
I'm not sure if there's a need for "required for postgresql".  Thoughts?

marvinirwin replied:
You're right, removed.  

I had it in there because the postgres connector will discover `pg_` tables (along with `public` tables) without a specified schema, on reconsideration, it's not worth warning about.

marvinirwin commented:
@raymondfeng  @dhmlau  @bajtos Got anything else?

dhmlau commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,49 @@
+`lb4 discover` command.
+
+**The LoopBack project must be built and contain the built datasource files in
+`PROJECT_DIR/dist/src/datasources.js**
```
I think there's a missing backtick.

marvinirwin replied:
fixedd

dhmlau commented:
@marvinirwin, thanks for updating the PR.  I've added one minor comment. 
It looks good to me in terms of the documentation.  We need to get the technical reviews from @bajtos @raymondfeng and team.  Thanks!

bajtos commented:
> Merge branch 'master' into pr-model-discovery

We don't merge branches and use rebase instead. That way the git history remains neatly linear. Could you please run `git rebase master` in your feature branch? See [How to rebase your branch](https://loopback.io/doc/en/contrib/code-contrib.html#how-to-rebase-your-branch)).

bajtos commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,49 @@
+`--all`: Skips the model prompt and discovers all of them
+
+`--outDir`: Specify the directory into which the `model.model.ts` files will be
+placed
```
I would expect `outDir` to default to `models`. Can we mention that in the docs and make sure the implementation does provide this default?

marvinirwin replied:
Added test for custom --outDir and mentioned default in docs

bajtos commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,49 @@
+`--schema`: Specify the schema which the datasource will find the models to
+discover
+
+`--schemaDefs`: The path to a `.json` file containing a model definition, like
```
Can we find a better name for this option, what that will convey the fact that we are loading models from JSON files instead of the database?

Few ideas to consider:

- `--import`
- `--import-definition`
- `--from-definition`
- `--from-file`
- `--from-lb3-models`

Also please make mention that the JSON file can contain either a single model definition object or an array with multiple definitions.

bajtos replied:
On the second thought, I am not sure if it's a good idea to mix `lb4 discover` with importing models from JSON definitions. I find it little bit confusing as a user, plus it makes the implementation more complex because of all those `if (importing from file) return` statements at the top of many generator steps.

Have you considered moving this functionality into `lb4 model` generator? I think it will give us CLI that feels pretty natural too.

```sh
lb4 model --import definition.json
# or perhaps
lb4 model --from-definition definition.json
```

marvinirwin replied:
Yes it's dramatically more intuitive, I'll move it to the model generator.  

Slightly related:  The `.json` loading was used to give me a way of testing the model discovery, which I still have a bit of trouble doing thoroughly.  A couple of solutions:

The memory datasource could be given `discoverModelDefinitions` and `discoverSchema`, but that might be out of the scope of this PR.  

Docker could also be used in the unit tests, but that definitely out of scope of this PR

marvinirwin replied:
@bajtos After looking around the code a bit I'm proposing keeping the --schemaDefs option.  And that I submit a seperate PR for adding the `--import` option to the model generator.

The current `discover --schemaDefs path` accepts an array of objects, which mimics how the `discover` command discovers multiple models.  I also need a way to mock data for the integration test.

bajtos replied:
Let's discuss testing and user experience independently.

For testing, I think you should mock `discoverModelDefinitions` and/or `discoverSchema` in such way that your tests sees the model schema you are interested in. I'll try to find time to create a code snippet showing how to mock those two methods.

For user experience, I still prefer importing via `lb4 model` over `--schemaDefs`. I am fine with leaving this option out of this initial pull request, as long as `--schemaDefs` is removed too.

bajtos replied:
I am thinking about the following solution for testing:

- Modify the datasource class used by your tests (`mem.datasource.js.txt`) to override discovery methods provided by `juggler.DataSource`.
- In these overridden methods, load schema from `test-schema-defs.json` and return the relevant parts depending on which discovery method was called.

A mock-up to illustrate what I mean:

```ts
export class MemDataSource extends juggler.DataSource {
  async discoverSchema(name, options) {
    await ensureSchemaDefsLoaded();
    const schema = this.schemas.find(s => s.name === name);
    if (schema) return schema;
    throw new Error(`Schema not found: ${name}`);
  }

  private schemas;
  private async ensureSchemaDefsLoaded() {
    if (schemas) return;
    const schemaFile = path.resolve(__dirname, 'test-schema-defs.json');
    _schemas = JSON.parse(await fs.readFile(schemaFile, 'utf-8'));
  }
}    
```

marvinirwin replied:
I'm pretty embarrassed I didn't think of implementing my own DataSource.  The testing is now much easier.    

https://github.com/strongloop/loopback-next/blob/5445c58c09760df39f5d010a136979f36c57b217/packages/cli/test/fixtures/discover/mem.datasource.js.txt

bajtos commented on docs/site/Model.md:
```diff
@@ -65,6 +65,11 @@ decorators for adding metadata to your TypeScript/JavaScript classes in order to
+## Model Discovery
+
+Some connectors support the ability to
+[Discover](Discovering-models-from-relational-databases.md) models.
```
Let's move this section _after_ "Definition of a Model" please.

Here is how I am thinking about the layout of this page:

- Overview
- Definition of a Model
- Model Discovery
- Using the Juggler Bridge

I am also proposing to slightly rephrase the text.

```markdown
LoopBack can automatically create model definitions by discovering the schema 
of your database. See 
[Discovering models](Discovering-models-from-relational-databases.md)
for more details and a list of connectors supporting model discovery.
```

marvinirwin replied:
done

bajtos commented on docs/site/Discovering-models-from-relational-databases.md:
```diff
@@ -0,0 +1,49 @@
+title: 'Discovering models from relational databases'
+keywords: LoopBack 4.0, LoopBack-Next
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Discovering-models-from-relational-databases.html
```
IIRC, Juggler supports several discovery modes. Besides parsing SQL schema, it can also define models based on example data. I think it would be great to implement support for this feature at some point in the future (but not in this pull request).

To keep our doors open, I am proposing to rename this page and leave out the part about relational databases.
- file name: `Discovering models.md`
- permalink: `/doc/en/lb4/Discovering-models.md`

Or perhaps also rename the page to "Model discovery" (`Model-discovery.md)? IDK.

marvinirwin replied:
done

bajtos commented on packages/cli/generators/discover/index.js:
```diff
@@ -0,0 +1,279 @@
+        'The path to a .json file with an array of schema definition',
+      default: undefined,
+    });
+  }
```
I don't see `--outDir` defined in generator options. It's defined in the docs though. What am I missing?

marvinirwin replied:
Woops, fixed

bajtos commented on packages/cli/lib/model-discoverer.js:
```diff
@@ -0,0 +1,94 @@
+      o[k] = JSON.stringify(o[k]);
+    }
+  });
+  // o.id could be 1 or 0, typescript will be angry if it is not a boolean
```
This is a bug in our type definitions. Can we fix the typedefs to correctly describe `id` as `boolean | number` please?

marvinirwin replied:
I'm not sure how to do this

bajtos replied:
I think we need to change the following line:

https://github.com/strongloop/loopback-next/blob/6f62444601faba789b798df953d14c2aababf26b/packages/repository/src/model.ts#L31

```diff
-  id?: boolean;
+  id?: boolean | number;
```

bajtos replied:
It may be best to submit this fix as a standalone pull request, that way we can get it landed quickly.

marvinirwin replied:
I've submitted it here https://github.com/strongloop/loopback-next/pull/2406

bajtos commented on packages/cli/package-lock.json:
```diff
@@ -0,0 +1,9825 @@
+{
+  "name": "@loopback/cli",
+  "version": "1.2.2",
+  "lockfileVersion": 1,
```
I believe we are intentionally not using package-lock files. Please remove it from the pull request.

marvinirwin replied:
fixed

bajtos replied:
I still see the package-lock file included in the PR. PTAL again.

bajtos commented on packages/cli/test/fixtures/discover/mem.datasource.js.txt:
```diff
@@ -0,0 +1,29 @@
+"use strict";
+var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
```
Hmm, keeping versioning JS files is not pretty. It makes me wonder how are tests for other generators (e.g. `lb4 repository`) dealing with datasource discovery? Could you PTAL and see if there is a way how to use an existing approach? Do we actually need this `.js` file? Isn't it enough to have datasource `.json` file?

marvinirwin replied:
You're right, I'm going to try and do this with the `legacy.Datasource` constructor and a JSON file.  The only downside I can see is that if any config is defined in`server/datasources/datasource.ts` it won't be loaded properly

marvinirwin replied:
I've tried this, but if I only use a .json file it will crash because the `lb4` command doesn't have the connector.  Is it worth it to try and bootstrap the connectors found in the discovering `node_modules`?

bajtos replied:
Could you please create a new feature branch in your repo where you will commit the code showing what have you tried, so that I can run that version myself and better understand what's going on under the hood? I am afraid I am not able to build a good picture in my head just from our discussion.

marvinirwin replied:
### Compare loading from `name.datasource.json`, to loading `name.datasource.js`

(requires lb4 project with non-memory datasource in `dist`, it doesn't have to connect successfully)

```
git clone https://github.com/marvinirwin/loopback-next --single-branch --branch discovering-from-json  .;
cd packages/cli;
npm run build;
npm link;
cd SOME_LB4_PROJECT;
//  new DataSource(JSON.parse(fs.readFileSync(pathToJson).toString()))
// This should tell you to npm install the connector
LOAD_JSON_DATASOURCES=true lb4 discover;
// new require(pathToJavascript);
// This will not tell you to npm install the connector
LOAD_JSON_DATASOURCES= lb4 discover;
```

bajtos commented on packages/cli/test/test-utils.js:
```diff
@@ -3,7 +3,9 @@
-'use strict';
+const fsextra = require('fs-extra');
+
+('use strict');
```
Please keep `use strict` at the top.

```js
'use strict';

const fsextra = require('fs-extra');
```

marvinirwin replied:
fixed

bajtos commented on packages/cli/test/test-utils.js:
```diff
@@ -124,8 +126,14 @@ exports.givenLBProject = function(rootDir, options) {
+          fs.writeFileSync(fullPath, theFile.content);
+        }
+      } catch (e) {
+        console.error(e);
```
When the file cannot be created then the test should fail on that error.

Please remove the `try`/`catch` block and let the test runner handle any error thrown by this helper.

marvinirwin replied:
fixed

bajtos commented:
@marvinirwin I see that you pushed new commits. Is the patch ready for another review? Most of us are not receiving notifications when a new commit is pushed into a pull request, please post a comment when it's time to review the changes again.

marvinirwin commented:
@bajtos  New fixes

1. mentioned exclusivity of --schemaDefs
2. added --import option to the model cli tool
3. renamed from Discovering-models-from-relational-databases.md to Discovering-models.md
4. moved Discovering Models below Definition of a model.

marvinirwin commented:
@bajtos New commit!

1. Removed `--import` and `--schemaDefs`
2. Testing `lb4 discover` is now done with a custom `DataSource`
3. Submitted https://github.com/strongloop/loopback-next/pull/2406

Royedc4 commented:
Hey guys, I've been a long term user of 2.X and this feature is important for us. How could we try this PR version to use the `lb4 discover`?

marvinirwin commented:
@Royedc4 

> Hey guys, I've been a long term user of 2.X and this feature is important for us.  How could we try this PR version to use the `lb4 discover`?

I created this PR and I'm waiting on @bajtos for feedback right now, but I use it often.

```
git clone -b pr-model-discovery --single-branch https://github.com/marvinirwin/loopback-next;
cd loopback-next/packages/cli;
npm install && npm run build && npm run link;
cd `LOOPBACK_4_PROJECT`;
npm run build;
lb4 discover;
```

bajtos commented on packages/cli/package.json:
```diff
@@ -48,6 +48,7 @@
     "fs-extra": "^7.0.1",
     "json5": "^2.0.1",
     "lodash": "^4.17.11",
+    "loopback-datasource-juggler": "^4.5.3",
```
IIUC, `loopback-datasource-juggler` is used only by the test suite. If that's true, then please move it from `dependencies` to `devDependencies`.

marvinirwin replied:
WOops, fixed

bajtos commented on packages/cli/test/fixtures/discover/index.js:
```diff
@@ -0,0 +1,19 @@
+    path: 'dist/src/datasources',
+    file: 'mem.datasource.json',
+    content: fs.readFileSync(require.resolve('./mem.datasource.json')),
+  },*/
```
Please remove commented-out code. Also if `test-schema-defs.json` and `mem.datasource.json` are no longer used, then remove those files too.

marvinirwin replied:
done

bajtos commented on docs/site/Discovering-models.md:
```diff
@@ -0,0 +1,45 @@
+`lb4 discover` command.
+
+**The LoopBack project must be built and contain the built datasource files in
+`PROJECT_DIR/dist/src/datasources.js`**
```
I don't think LB4 projects have a file called `dist/src/datasources.js`. Did you mean `dist/src/datasources/*.js` instead?

Also please note that we have recently changed the organization of project files in the `dist` folder, the compiler is configured to compile `src/*` to `dist/*`, i.e. `src/datasources/db.datasource.ts` is compiled to `dist/datasources/db.datasource.js` now.

Please update the README (and the implementation?) accordingly.

marvinirwin replied:
Yes, that's what I meant.  Fixed

bajtos commented on packages/cli/lib/model-discoverer.js:
```diff
@@ -0,0 +1,105 @@
+/**
+ * Returns the entire model definition for a model
+ */
+async function discoverSingleModel(ds, options, name) {
```
```suggestion
async function discoverSingleModel(ds, modelName, options) {
```

`options` is usually the last argument. I find the argument name `name` as too generic, let's use `modelName` to make it obvious what kind of a name we are expecting.

marvinirwin replied:
done

bajtos commented on packages/cli/lib/model-discoverer.js:
```diff
@@ -0,0 +1,105 @@
+  );
+}
+
+const DEFAULT_DATASOURCE_DIRECTORY = './dist/src/datasources';
```
```suggestion
const DEFAULT_DATASOURCE_DIRECTORY = './dist/datasources';
```

I thought we already have code to discover datasources elsewhere in the CLI codebase, isn't such functionality required by `lb4 repository` command? Can we find a way how to extract a shared helper and avoid code duplication?

marvinirwin replied:
done

bajtos commented on packages/cli/lib/model-discoverer.js:
```diff
@@ -0,0 +1,105 @@
+  });
+  // o.id could be 1 or 0, typescript will be angry if it is not a boolean
+  if (o.id) {
+    o.id = !!o.id;
```
#2406 has been landed, I think we can remove this workaround now?

```diff
- if (o.id) {
-   o.id = !!o.id;
- }
```

marvinirwin replied:
done

bajtos commented on packages/cli/lib/model-discoverer.js:
```diff
@@ -0,0 +1,105 @@
+  o.tsType = o.type;
+};
+
+function getAllDataSourceFiles(dir = DEFAULT_DATASOURCE_DIRECTORY) {
```
Here is an existing implementation for looking up all datasources. Please re-use the same low-level building blocks instead of creating new ones.

https://github.com/strongloop/loopback-next/blob/71292e9ac1b3d89ebfe284a659264cbb83dbe814/packages/cli/generators/repository/index.js#L232-L310

I think it would be great to extract `promptDataSourceName` generator action into a shared helper that can be called from multiple generators. 

Example usage of such helper in `lb4 datasource`:

```ts
const filter = item => {
    debug(`data source inspecting item: ${item}`);
    const result = utils.isConnectorOfType(
      VALID_CONNECTORS_FOR_REPOSITORY,
      this.artifactInfo.datasourcesDir,
      item,
    );
    return result !== false;
  });
};

return promptDataSourceName(this.options.datasource, this.artifactInfo, filter)
  .then({dataSourceClass} => {
    Object.assign(this.artifactInfo, {dataSourceClass});
    })
  .catch(err => {
    debug(`Error during datasource prompt: ${err}`);
    return this.exit(err);
  });
```

bajtos replied:
> I think it would be great to extract promptDataSourceName generator action into a shared helper that can be called from multiple generators.

Feel free to leave out this refactoring out of scope of this pull request if it's easier that way.

marvinirwin replied:
I'll do this in another PR

marvinirwin replied:
> Here is an existing implementation for looking up all datasources. Please re-use the same low-level building blocks instead of creating new ones.

Fixed, I use [getArtifactList](https://github.com/strongloop/loopback-next/blob/master/packages/cli/lib/utils.js#L211) now.  I'll make the general prompt in another PR

bajtos replied:
I am afraid you addressed my comment partially. It's good to see that generators are using the shared `getArtifactList` function now, but we need to change other places like `loadDataSourceByName` too, so that `getAllDataSourceFiles` and `DEFAULT_DATASOURCE_DIRECTORY` can be removed completely.

My concern is that right now, the convention for layout of datasource files is duplicated in multiple places and when we change it in the future, we won't update all of these places. 

Anyhow, I am ok to leave further improvements in this area out of scope of this initial pull request.

marvinirwin commented:
@bajtos  I've made your changes, now waiting for @jannyHou, @marioestradarosa , @raymondfeng

raymondfeng commented:
@marvinirwin Thank you for the contribution! I'm rebasing the PR against latest master and squashing them into one commit. Will push back to your branch if it's permitted.

raymondfeng commented:
@marvinirwin I have pushed my changes back to your branch. It is up to date with upstream master and fixes an issue related to `dist` (not more `dist/src`).

bajtos commented on packages/cli/test/test-utils.js:
```diff
@@ -2,9 +2,9 @@
-
 'use strict';
 
+const fsextra = require('fs-extra');
```
AFAIK, `fsextra` is exposing all `fs` APIs and adds some new ones. There is no need to have both `fs` and `fsextra` variables. Here is what we usually do:

```js
- const fs = require('fs');
+ const fs = require('fs-extra');
```

bajtos replied:
> AFAIK, `fsextra` is exposing all `fs` APIs and adds some new ones. There is no need to have both `fs` and `fsextra` variables.

@marvinirwin PTAL ☝️

marvinirwin replied:
woops, fixed

marvinirwin commented:
@bajtos  I think I understand what you mean about the duplication of logic in loading datasources now.  

`DEFAULT_DATASOURCE_DIRECTORY` will be replaced with `utils.distRootDir = 'dist'`.  
`loadDataSourceByName` will be replaced with loading via path with [dataSourceToJSONFilename] (https://github.com/strongloop/loopback-next/blob/861256c4e893f6b26e1d2b81d69e614ab5de9e9f/packages/cli/lib/utils.js#L507)

Once these changes are implemented both the `Repository` and `Discovery` generator can create their datasource prompts in the exact same way.

 I'll begin that PR once this one has been merged.

bajtos commented:
@marvinirwin 

> I'll begin that PR once this one has been merged.

Sounds good.

I noticed that our code coverage dropped by 0.4% with your patch in place, which is more than we usually accept.

I reviewed the code coverage data, it looks like you pull request is missing tests for datasource prompt when discovering a single model (see e.g. https://coveralls.io/builds/22183916/source?filename=packages/cli/generators/discover/index.js)) and for the code mapping discovered types to TypeScript snippets (see https://coveralls.io/builds/22183916/source?filename=packages/cli/generators/model/index.js#L457).

How easy or difficult it is to write tests covering these two areas?

marvinirwin commented:
@bajtos Thanks for reminding me about the discovering single models.  I added tests for that and some tests ensuring nicer error messages.    

I'm not sure about the loopback->typescript mapping parts.  I think I can hit them all in a more exhaustive test, but I think that's out of the scope of this PR.

bajtos commented on packages/cli/test/integration/generators/discover.integration.js:
```diff
@@ -10,6 +10,7 @@ const path = require('path');
 const testlab = require('@loopback/testlab');
 
+const expect = testlab.expect;
 const TestSandbox = testlab.TestSandbox;
```
Nitpick: these two lines can be simplified as follows:

```js
const {expect, TestSanbox} = testlab;
```

bajtos commented on packages/cli/test/fixtures/discover/mem.datasource.js.txt:
```diff
@@ -0,0 +1,106 @@
+const DataSource = require('loopback-datasource-juggler').DataSource;
```
I would prefer to avoid direct dependency on loopback-datasource-juggler and import `DataSource` class constructor from `@loopback/repository`.

Not a big deal, we can improve this in a follow-up pull request.

bajtos commented on packages/cli/test/fixtures/discover/mem.datasource.js.txt:
```diff
@@ -0,0 +1,106 @@
+  },
+];
+
+// TODO add .txt to this extension so it doesn't get called by global tests`
```
This TODO does not seem to be relevant any more, the file already has `.txt` extension.

marvinirwin commented on docs/site/Discovering-models.md:
```diff
@@ -0,0 +1,45 @@
+LoopBack makes it simple to create models from an existing relational database.
+This process is called _discovery_ and is supported by the following connectors:
+
+- [Cassandra](Cassandra-connector.html)
```
Huh, I forgot about those.  From a cursory look, I'm not sure if I can relative paths with those.

dhmlau replied:
@marvinirwin , I believe you're getting the list from the LB3 page: https://loopback.io/doc/en/lb3/Discovering-models-from-relational-databases.html.  We have [a work item ](https://github.com/strongloop/loopback-next/issues/2598) to add those connector docs pages back.  
For this PR, I'd like to propose to simply remove the links of the connectors.  When we get to https://github.com/strongloop/loopback-next/issues/2598, we'll add back the links.

marvinirwin commented on packages/cli/generators/discover/index.js:
```diff
@@ -0,0 +1,256 @@
+
+  /**
+   * If we have a dataSource, attempt to load it
+   * If we have a schemaDef path, attempt to load it
```
good catch, `schemaDefs` had been removed

bajtos commented:
@marvinirwin could you please address the comments raised by @nabdelgadir above, and also add this new command to the [CLI README](https://github.com/strongloop/loopback-next/blob/master/packages/cli/README.md)? I'd like to get them fixed before we land this patch.

marvinirwin commented:
@bajtos done

raymondfeng commented:
> add this new command to the CLI README?

You can run `packages/cli/bin/cli-main.js discover --help` to print out the command help that can be added to README.

marvinirwin commented:
@raymondfeng  

The first line of my output is not correct, but I'm not sure how to change it.  (discover does not have a name parameter)

```
Usage:
  lb4 discover [<name>] [options]

Options:
  -h,    --help           # Print the generator's options and usage
         --skip-cache     # Do not remember prompt answers                                              Default: false
         --skip-install   # Do not automatically install dependencies                                   Default: false
         --force-install  # Fail on install dependencies error                                          Default: false
  -c,    --config         # JSON file name or value to configure options
  -y,    --yes            # Skip all confirmation prompts with default or provided value
         --format         # Format generated code using npm run lint:fix
  -ds,   --dataSource     # The name of the datasource to discover
         --views          # Boolean to discover views                                                   Default: true
         --schema         # Schema to discover
         --all            # Discover all models without prompting users to select                       Default: false
         --outDir         # Specify the directory into which the `model.model.ts` files will be placed

Arguments:
  name  # Name for the discover  Type: String  Required: false
```

nabdelgadir commented on packages/cli/README.md:
````diff
@@ -251,7 +251,27 @@ Run the following command to install the CLI.
+
+    ```
+
+11) To list available commands
````
Can you change this to `11.` instead of `11)` (same for `12.`)?

marvinirwin replied:
nice catch

bajtos commented:
> The first line of my output is not correct, but I'm not sure how to change it. (discover does not have a name parameter)

@raymondfeng could you please help here? I think you now the internal CLI architecture best.

Personally, I can live with incorrect help message for the time being, I see much more value in having the command `lb4 discover` available to all users.

marvinirwin commented:
@bajtos Alright I've put the mildly incorrect output from `lb4 discover --help` in the readme.  The Travis is failing right now, though I'm not sure if it's my fault or not

bajtos commented:
> The Travis is failing right now, though I'm not sure if it's my fault or not

See https://travis-ci.org/strongloop/loopback-next/jobs/516258539. Your feature branch seems to contain a bunch of commits that are violating our [Commit Message Guidelines](https://loopback.io/doc/en/contrib/code-contrib-lb4.html#commit-message-guidelines).

Could you please try to rebase your feature branch on top of the latest master? See https://loopback.io/doc/en/lb4/submitting_a_pr.html#6-rebasing

We can help you with this last step if you like, just let us know.

marvinirwin commented:
@bajtos Done, I think

dhmlau commented:
@slnode test please

b-admike commented:
@marvinirwin I've rebased your branch and resolved some conflicts, PTAL and if you are satisfied, we can land your patch.

marvinirwin commented:
@b-admike 

LGTM

Royedc4 commented:
Awwwwwesome, Finally 🎉
When will this be available in loopback for everyone?

dhmlau commented:
There're some changes we want to go in for the same release.  Once that's approved, we'll release.  Hopefully very soon 🤞
